parsers/ZA.py

```diff
@@ -18,118 +55,77 @@ def get_url() -> str:
-        3: "ignored",  # Hydro_Water_SCO_Pumping            Probably electricity consumed by the plant itself (even) when not generating power. Can be ignored.
-        4: "hydro",  # Pumped_Water_SCO_Pumping
-        5: "ignored",  # Thermal_Generation                 sum of 0, 1, 2, 3, 4. Can be ignored.
-        6: "nuclear",  # Nuclear_Generation
-        7: "ignored",  # International_Imports
-        8: "oil",  # Eskom_OCGT_Generation
-        9: "gas",  # Eskom_Gas_Generation
-        10: "oil",  # Dispatchable_IPP_OCGT
-        11: "hydro",  # Hydro_Water_Generation
-        12: "hydro",  # Pumped_Water_Generation
-        13: "ignored",  # IOS_Excl_ILS_and_MLR              Interruption of Supply. Can be ignored.
-        14: "ignored",  # ILS_Usage                         Interruptible Load Shed = companies paid not to consume electricity. Can be ignored.
-        15: "ignored",  # Manual_Load_Reduction_MLR         MLS = forced load shedding. Can be ignored.
-        16: "wind",  # Wind
-        17: "solar",  # PV
-        18: "solar",  # CSP
-        19: "unknown",  # Other_RE
-    }
-
-    all_data = []
-
-    for _, (datetime_key, prod_data) in enumerate(production_csv_data.items()):
-
-        # The production values (MW) should never be negative. Use None, or omit the key if # a specific production mode is not known.
-        # ---
-        # storage values can be both positive (when storing energy) or negative (when the
-        # storage is discharged).
-
-        data = {
-            "zoneKey": zone_key,
-            "datetime": datetime_key.datetime,
-            "production": {
-                "coal": 0.0,
-                "gas": 0.0,
-                "hydro": 0.0,
-                "nuclear": 0.0,
-                "oil": 0.0,
-                "solar": 0.0,
-                "wind": 0.0,
-                "unknown": 0.0,
-            },
-            "storage": {"hydro": 0.0},
-            "source": " https://www.eskom.co.za",
-        }
-
-        storage_inversion_idcs = [4, 12]
-        production_idcs = [0, 6, 8, 9, 10, 11, 16, 17, 18, 19]
-
-        # Ignored values
-        # 1, 2, 3, 5, 7, 13, 14, 15
-        # TODO:
-        # - 7 (international imports) can be further implemented in exchange function.
-
-        for j, prod_data_value in enumerate(prod_data):
-            if j in storage_inversion_idcs:
-                data["storage"][column_mapping[j]] = round(
-                    data["storage"][column_mapping[j]] + float(prod_data_value) * -1,
-                    13,
-                )
-            elif j in production_idcs:
-                data["production"][column_mapping[j]] = round(
-                    data["production"][column_mapping[j]] + float(prod_data_value),
-                    13,
-                )
-
-        all_data.append(data)
-
-    return all_data
+    if not res.ok:
+        raise ParserException(
+            "ZA.py",
```

Comment:
Not related to this PR, but couldn't we infer the filename in a smarter way? I also see that we in some places use `zone_key` instead of the filename - so maybe a general approach could help clean that up too :)

Reply:
For the parser exception we log the parser, message and zone_key. But yeah hopefully we can make this more "automatic". I think Robin has improving logging on some roadmap somewhere 😉